src/app/components/todos/todos.component.ts

```diff
@@ -2,32 +2,82 @@ import { Component } from '@angular/core';
 import { Store } from '@ngrx/store';
 import { Observable } from 'rxjs/Observable';
 import 'rxjs/add/operator/map';
+import 'rxjs/add/observable/combineLatest';
 
 import { TodosActions } from '../../actions/todos.actions';
-import { TodosState } from '../../reducers/todos.reducer';
 import { Todo } from '../../models/todo';
+import { TodosState } from '../../reducers/todos.reducer';
+import { FiltersState } from '../../reducers/filters.reduser';
 
 @Component({
   selector: 'app-todos',
   templateUrl: './todos.component.html'
 })
 export class TodosComponent {
   textTodo: string;
-  todoItems: Observable<Todo[]>;
+  editableTodo: Todo = null;
+  todos$: Observable<Todo[]>;
+  activeFilter$: Observable<string>;
 
   constructor(private store: Store<TodosState>, private actions: TodosActions) {
-    const todos: Observable<TodosState> = this.store.select('todos');
-
     this.textTodo = '';
-    this.todoItems = todos.map(tds => tds.todoItems);
+
+    const state$: Observable<{todos: TodosState, filters: FiltersState}> = Observable.combineLatest(
+      this.store.select('todos'),
+      this.store.select('filters'),
+      (todos, filters) => ({todos, filters})
+    );
+
+    this.todos$ = state$.map((st: {todos: TodosState, filters: FiltersState}) => {
+      this.editableTodo = st.todos.editableTodo;
+      return st.filters.filter(st.todos.todoItems);
+    });
+
+    this.activeFilter$ = state$.map(st => st.filters.activeFilter);
     this.store.dispatch(this.actions.getTodos());
   }
 
   submitTodo(evt: Event): void {
     evt.preventDefault();
     if (this.textTodo) {
-      this.store.dispatch(this.actions.addTodo(this.textTodo));
+      if (this.editableTodo) {
+        this.store.dispatch(this.actions.updateTodo(this.textTodo));
+      } else {
+        this.store.dispatch(this.actions.addTodo(this.textTodo));
+      }
+
       this.textTodo = '';
     }
   }
+
+  removeTodo(todo: Todo) {
+    this.store.dispatch(this.actions.removeTodo(todo));
+  }
+
+  toggleTodo(todo: Todo) {
+    this.store.dispatch(this.actions.toggleTodo(todo));
+  }
+
+  editTodo(todo: Todo): void {
+    this.store.dispatch(this.actions.editTodo(todo));
+    this.textTodo = this.editableTodo ? todo.text : '';
+  }
+
+  showAllTodos(filter: string): void {
```

Comment:
those 3 methods are the same - move logic from template to here and use ngFor in the template to generate a list of filters applied